x-pack/legacy/plugins/siem/server/lib/detection_engine/routes/rules/utils.test.ts

```diff
@@ -1224,4 +1230,96 @@ describe('utils', () => {
+      const ndJsonStream = new Readable({
+        read() {
+          this.push(multipartPayload);
+          this.push(null);
+        },
+      });
+      const rulesObjectsStream = createRulesStreamFromNdJson(ndJsonStream, 1000);
+      const parsedObjects = await createPromiseFromStreams<PromiseFromStreams[]>([
+        rulesObjectsStream,
+      ]);
+      const [errors, output] = getTupleDuplicateErrorsAndUniqueRules(parsedObjects, false);
+      const isInstanceOfError = output[0] instanceof Error;
+
+      expect(isInstanceOfError).toEqual(true);
+      expect(errors.length).toEqual(0);
+    });
+
+    test('returns tuple of duplicate conflict error and single rule when rules with matching rule-ids passed in and `overwrite` is false', async () => {
+      const rule = getSimpleRule('rule-1');
+      const rule2 = getSimpleRule('rule-1');
+      const ndJsonStream = new Readable({
+        read() {
+          this.push(`${JSON.stringify(rule)}`);
+          this.push('\n');
+          this.push(`${JSON.stringify(rule2)}`);
+          this.push(null);
+        },
+      });
+      const rulesObjectsStream = createRulesStreamFromNdJson(ndJsonStream, 1000);
+      const parsedObjects = await createPromiseFromStreams<PromiseFromStreams[]>([
+        rulesObjectsStream,
+      ]);
+      const [errors, output] = getTupleDuplicateErrorsAndUniqueRules(parsedObjects, false);
+
+      expect(output.length).toEqual(1);
+      expect(errors).toEqual([
+        {
+          error: {
+            message: 'More than one rule with rule-id: "rule-1" found',
+            status_code: 400,
+          },
+          rule_id: 'rule-1',
+        },
+      ]);
+    });
+
+    test('returns tuple of empty duplicate errors array and single rule when rules with matching rule-ids passed in and `overwrite` is true', async () => {
+      const rule = getSimpleRule('rule-1');
+      const rule2 = getSimpleRule('rule-1');
+      const ndJsonStream = new Readable({
+        read() {
+          this.push(`${JSON.stringify(rule)}`);
+          this.push('\n');
+          this.push(`${JSON.stringify(rule2)}`);
+          this.push(null);
+        },
+      });
+      const rulesObjectsStream = createRulesStreamFromNdJson(ndJsonStream, 1000);
+      const parsedObjects = await createPromiseFromStreams<PromiseFromStreams[]>([
+        rulesObjectsStream,
+      ]);
+      const [errors, output] = getTupleDuplicateErrorsAndUniqueRules(parsedObjects, true);
+
+      expect(output.length).toEqual(1);
+      expect(errors.length).toEqual(0);
+    });
+
+    test('returns tuple of empty duplicate errors array and single rule when rules without a rule-id is passed in', async () => {
+      const multipartPayload =
+        '{"name":"Simple Rule Query","description":"Simple Rule Query","risk_score":1,"rule_id":"","severity":"high","type":"query","query":"user.name: root or user.name: admin"}\n';
```

Comment:
Less ceremony and more types if you do this:

```ts
      const simpleRule = getSimpleRule('');
      const multipartPayloadStringify = JSON.stringify(simpleRule);
      const ndJsonStream = new Readable({
        read() {
          this.push(multipartPayloadStringify);
          this.push(null);
        },
      });
```

Kind of the niceness of that little helper function, `getSimpleRule`. However, this is technically just changing `rule_id` to be an empty string. If you wanted to instead test having it deleted then it would be:

```ts
      const simpleRule = getSimpleRule();
      delete simpleRule.rule_id;
      const multipartPayloadStringify = JSON.stringify(simpleRule);
      const ndJsonStream = new Readable({
        read() {
          this.push(multipartPayloadStringify);
          this.push(null);
        },
      });
```

I would update the test to reflect it either intends to pass in an empty string for `rule_id` or change the code to actually delete the `rule_id`?

Reply:
Thanks, that makes sense. I used the latter - deleting rule_id - as that was what I intended to test.